\begin{table*}
    \begin{tabular}{rrrrr}
        \toprule
        \#seqs & pggb & vg & minigraph-cactus & pfg \\
        \midrule
        4   & 0m 12s & 0m 1s  & 3m 47s  & 0m 1s \\
        16  & 0m 19s & 0m 1s  & 6m 30s  & 0m 1s \\
        64  & 1m 20s & 0m 2s  & 13m 30s & 0m 1s \\
        256 & 7m 20s & 0m 55s & 41m 49s & 0m 3s \\
        \bottomrule
    \end{tabular}
    \caption{
        Comparison of running times of pangenomic graph construction tools.
        Wall clock times are measured by the built-in bash command \texttt{time} and rounded up to the nearest second.
        In addition to the measured times, \texttt{pggb} and \texttt{vg} need preprocessing of the pangenome.
        \texttt{pggb} requires the fasta file to be indexed.
        This time was negligible in our case, but it can have a more significant impact when a more extensive dataset is considered.
        \texttt{vg} builds a pangenome from a VCF file or a multiple sequence alignment.
        In both cases, acquiring this data requires time-consuming steps.
        Furthermore, variant calling using a linear reference can introduce a reference bias.
        \label{tab:comparison}
    }
\end{table*}

